core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -0,0 +1,314 @@
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.api
+
+import java.util.Properties
+
+import kafka.integration.KafkaServerTestHarness
+import kafka.server.KafkaConfig
+import kafka.utils.TestUtils
+import org.apache.kafka.clients.consumer.{ConsumerConfig, ConsumerRecord, KafkaConsumer, OffsetAndMetadata}
+import org.apache.kafka.clients.producer.{ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.{KafkaException, TopicPartition}
+import org.apache.kafka.common.errors.ProducerFencedException
+import org.apache.kafka.common.protocol.SecurityProtocol
+import org.junit.{After, Before, Ignore, Test}
+import org.junit.Assert._
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.util.Random
+
+class TransactionsTest extends KafkaServerTestHarness {
+  val numServers = 3
+  val topic1 = "topic1"
+  val topic2 = "topic2"
+
+  override def generateConfigs : Seq[KafkaConfig] = {
+    TestUtils.createBrokerConfigs(numServers, zkConnect, true).map(KafkaConfig.fromProps(_, serverProps()))
+  }
+
+  @Before
+  override def setUp(): Unit = {
+    super.setUp()
+    val numPartitions = 3
+    val topicConfig = new Properties();
+    topicConfig.put(KafkaConfig.MinInSyncReplicasProp, 2.toString)
+    TestUtils.createTopic(zkUtils, topic1, numPartitions, numServers, servers, topicConfig)
+    TestUtils.createTopic(zkUtils, topic2, numPartitions, numServers, servers, topicConfig)
+  }
+
+  @After
+  override def tearDown(): Unit = {
+    super.tearDown()
+  }
+
+  @Test
+  def testBasicTransactions() = {
+    val producer = transactionalProducer("my-hello-world-transactional-id")
+    val consumer = transactionalConsumer()
+    try {
+      producer.initTransactions()
+
+      producer.beginTransaction()
+      producer.send(producerRecord(topic1, "1", "1", willBeCommitted = true))
+      producer.send(producerRecord(topic2, "3", "3", willBeCommitted = true))
+      producer.commitTransaction()
+
+      producer.beginTransaction()
+      producer.send(producerRecord(topic2, "2", "2", willBeCommitted = false))
+      producer.send(producerRecord(topic1, "4", "4", willBeCommitted = false))
+      producer.abortTransaction()
+
+      consumer.subscribe(List(topic1, topic2))
+
+      val records = pollUntilExactlyNumRecords(consumer, 2)
```

Comment:
How can we be sure that the aborted records wouldn't be returned if we waited a little longer? Maybe we should do the aborted transaction first to ensure that it. Also, it might make sense to have a second READ_UNCOMMITTED consumer and verify that all records are returned.